board/main_comms.h

```diff
@@ -47,106 +47,122 @@ int get_rtc_pkt(void *dat) {
   return sizeof(t);
 }
 
+// CAN comms state
 typedef struct {
   uint32_t ptr;
   uint32_t tail_size;
   uint8_t data[72];
-  uint8_t counter;
 } asm_buffer;
+asm_buffer can_read_overflow = {.ptr = 0U, .tail_size = 0U};
+asm_buffer can_write_overflow = {.ptr = 0U, .tail_size = 0U};
 
-asm_buffer can_read_buffer = {.ptr = 0U, .tail_size = 0U, .counter = 0U};
-uint32_t total_rx_size = 0U;
+bool add_magic = true;
 
+// for CAN read and send, we write out a buffer consisting of:
+// CAN_TRANSACTION_MAGIC + (N * CANPacket_t)
+// comms_can_read will output this buffer into chunks of size max_len.
+this buffer is split into chunks output by the comms_can_xxx functions.
 int comms_can_read(uint8_t *data, uint32_t max_len) {
-  uint32_t pos = 1;
-  data[0] = can_read_buffer.counter;
+  uint32_t pos = 0U;  // count of bytes written
+  bool added_magic = false;
+
+  // Start of a transaction
+  if (add_magic && (max_len >= sizeof(uint32_t))) {
+    *((uint32_t *)(void *) &data[0]) = CAN_TRANSACTION_MAGIC;
+    pos += sizeof(uint32_t);
+    add_magic = false;
+    can_read_overflow.ptr = 0U;
+    added_magic = true;
+  }
+
   // Send tail of previous message if it is in buffer
-  if (can_read_buffer.ptr > 0U) {
-    if (can_read_buffer.ptr <= 63U) {
-      (void)memcpy(&data[pos], can_read_buffer.data, can_read_buffer.ptr);
-      pos += can_read_buffer.ptr;
-      can_read_buffer.ptr = 0U;
-    } else {
-      (void)memcpy(&data[pos], can_read_buffer.data, 63U);
-      can_read_buffer.ptr = can_read_buffer.ptr - 63U;
-      (void)memcpy(can_read_buffer.data, &can_read_buffer.data[63], can_read_buffer.ptr);
-      pos += 63U;
-    }
+  if (can_read_overflow.ptr > 0U) {
+    uint32_t overflow_len = MIN(max_len - pos, can_read_overflow.ptr);
+    (void)memcpy(&data[pos], can_read_overflow.data, overflow_len);
+    pos += overflow_len;
+    (void)memcpy(can_read_overflow.data, &can_read_overflow.data[overflow_len], can_read_overflow.ptr - overflow_len);
+    can_read_overflow.ptr -= overflow_len;
   }
 
-  if (total_rx_size > MAX_EP1_CHUNK_PER_BULK_TRANSFER) {
```

Comment:
any reason to keep `MAX_EP1_CHUNK_PER_BULK_TRANSFER` without the counter?

Reply:
not sure about the history of this PR now, didn't I already remove the counter and this limit in the first commit?

Reply:
oh seems like I had to add it back

Reply:
thought it was pretty stable where I left it?